File: benchmarks/tsb/bench_swaplevel_dataframe.ts
/**
 * Benchmark: swapLevelDataFrame / reorderLevelsDataFrame on 50k-row MultiIndex DataFrame.
 *
 * Mirrors pandas:
 *   - DataFrame.swaplevel(i, j, axis=0)     → swapLevelDataFrame
 *   - DataFrame.reorder_levels(order, axis=0) → reorderLevelsDataFrame
 *
 * Dataset: 50 000-row × 3-column DataFrame with a 3-level MultiIndex row index.
 *
 * Outputs JSON: {"function": "swaplevel_dataframe", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, MultiIndex, swapLevelDataFrame, reorderLevelsDataFrame } from "../../src/index.ts";

const N = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

const levA = Array.from({ length: N }, (_, i) => `a${i % 100}`);
const levB = Array.from({ length: N }, (_, i) => i % 500);
const levC = Array.from({ length: N }, (_, i) => i % 10);
const tuples: [string, number, number][] = levA.map((v, i) => [v, levB[i], levC[i]]);
const idx = new MultiIndex({ tuples });

const df = DataFrame.fromColumns(
  {
    x: Array.from({ length: N }, (_, i) => i * 1.0),
    y: Array.from({ length: N }, (_, i) => i * 2.0),
    z: Array.from({ length: N }, (_, i) => i * 3.0),
  },
  { index: idx },
);

for (let i = 0; i < WARMUP; i++) {
  swapLevelDataFrame(df, 0, 1);
  swapLevelDataFrame(df, 0, 2);
  reorderLevelsDataFrame(df, [2, 0, 1]);
  reorderLevelsDataFrame(df, [1, 2, 0]);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  swapLevelDataFrame(df, 0, 1);
  swapLevelDataFrame(df, 0, 2);
  reorderLevelsDataFrame(df, [2, 0, 1]);
  reorderLevelsDataFrame(df, [1, 2, 0]);
}
const total_ms = performance.now() - start;

console.log(
  JSON.stringify({
    function: "swaplevel_dataframe",
    mean_ms: total_ms / ITERATIONS,
    iterations: ITERATIONS,
    total_ms,
  }),
);

File: benchmarks/pandas/bench_swaplevel_dataframe.py
"""
Benchmark: DataFrame.swaplevel / DataFrame.reorder_levels on 50k-row MultiIndex DataFrame.

Mirrors tsb bench_swaplevel_dataframe.ts.

Dataset: 50 000-row × 3-column DataFrame with a 3-level MultiIndex row index.

Outputs JSON: {"function": "swaplevel_dataframe", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time

import numpy as np
import pandas as pd

N = 50_000
WARMUP = 5
ITERATIONS = 30

lev_a = [f"a{i % 100}" for i in range(N)]
lev_b = [i % 500 for i in range(N)]
lev_c = [i % 10 for i in range(N)]

idx = pd.MultiIndex.from_arrays([lev_a, lev_b, lev_c], names=["L0", "L1", "L2"])
df = pd.DataFrame(
    {
        "x": np.arange(N, dtype=float),
        "y": np.arange(N, dtype=float) * 2.0,
        "z": np.arange(N, dtype=float) * 3.0,
    },
    index=idx,
)

for _ in range(WARMUP):
    df.swaplevel(0, 1, axis=0)
    df.swaplevel(0, 2, axis=0)
    df.reorder_levels([2, 0, 1], axis=0)
    df.reorder_levels([1, 2, 0], axis=0)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.swaplevel(0, 1, axis=0)
    df.swaplevel(0, 2, axis=0)
    df.reorder_levels([2, 0, 1], axis=0)
    df.reorder_levels([1, 2, 0], axis=0)
total_ms = (time.perf_counter() - start) * 1000

print(
    json.dumps(
        {
            "function": "swaplevel_dataframe",
            "mean_ms": total_ms / ITERATIONS,
            "iterations": ITERATIONS,
            "total_ms": total_ms,
        }
    )
)
